### ISSUE TITLE
Play, pause and volume message

### ISSUE THREAD

**Issue by salamay** [jonathantribouharet/flutter_cast#28]
Please, i have searched on how to play, pause and stop chromecast but none is working,  what is the right message to send in json format to play, pause and stop

**Comment by kayosilva** [comment]
I'm having the same problem, I've searched everywhere how to send the play/pause event to chromecast and I didn't find it.
I've already tried to send the connect event with the "autoPlay":false attribute, but that doesn't work either.

**Comment by JoseBarreto1** [comment]
I also have that problem

**Comment by Tiagocf2** [comment]
I found this link on issue #22 about the messages: https://developers.google.com/cast/docs/media/messages#Play

Also check: https://github.com/jonathantribouharet/flutter_cast/issues/8#issuecomment-1001748664

**Comment by Tiagocf2** [comment]
I only managed to make the PLAY and PAUSE message work though, for some reason my VOLUME message is not being accepted...

<details open>
<summary><b>My message</b></summary>

```dart
session.sendMessage(
  CastSession.kNamespaceMedia, {
    'type': 'VOLUME',
    'volume': {'muted': true},
    'mediaSessionId': 1,
  }
);
```
</details>
<details open>
<summary><b>Response message</b></summary>

```dart
{requestId: 107, type: INVALID_REQUEST, reason: INVALID_COMMAND}
```
</details>

**Comment by kayosilva** [comment]
Based on the information @JoseBarreto1  provided.

When you connect to chromecast you will get the event response and you will be able to get the **requestID**.

After the message with the media data you sent to chromecast, you will also receive a media playback status event, at which time you will be able to get the **mediaSessionId**.

These two attributes will be useful to send an event to control SEEK, PLAY/PAUSE to chromecast. In the example implementation of the plugin page https://pub.dev/packages/cast/example I made some changes to achieve it, as follows:

```
// ...
class _MyHomePageState extends State<MyHomePage> {
  Future<List<CastDevice>>? _future;
  late dynamic requestId;
  late dynamic mediaSessionId;
  late CastSession session;

@override
  void initState() {
// ...


Future<void> _connectAndPlayMedia(BuildContext context, CastDevice object) async {
    session = await CastSessionManager().startSession(object);

// ...
    var index = 0;
    session.messageStream.listen((message) {
       index += 1;

       if (index == 2) {
         Future.delayed(Duration(seconds: 5)).then((x) {
            requestId = message['requestId'];
            _sendMessagePlayVideo(session);
         });
       }
   
       if (message['status'] != null && message['status'][0] != null && message['status'][0]['playerState'] == 'PLAYING') {
            mediaSessionId = message['status'][0]['mediaSessionId'];
       }
    });

//...   

}

// ...

pause() {
    Map<String, dynamic> action = {
      'type': 'PAUSE',
      'mediaSessionId': mediaSessionId,
      'requestId': requestId
    };
    this.isPlaying = false;
    session.sendMessage(CastSession.kNamespaceMedia, action);
  }

  play() {
    Map<String, dynamic> action = {
      'type': 'PLAY',
      'mediaSessionId': mediaSessionId,
      'requestId': requestId
    };
    this.isPlaying = true;
    session.sendMessage(CastSession.kNamespaceMedia, action);
  }

  seek(seconds) {
    Map<String, dynamic> action = {
      'type': 'SEEK',
      'mediaSessionId': mediaSessionId,
      'requestId': requestId,
      'currentTime': seconds
    };
    session.sendMessage(CastSession.kNamespaceMedia, action);
  }

// ...


```

**Comment by guyluz11** [comment]
Hey all.
It looks like there is a couple of issues about people not sure how to use features or asking if the feature exists and this (in my opinion) underlines a problem with the package and code.


First of all, we should 
1. create an enum for 'type' for clear code (maybe with the option of passing a string if needed).
2. Easy functions for play, pause, volume...
3. And update the documentation to be more precise and contain links to pub.dev page, link Google documentation of casting as described in other issue #22 .

It looks like the repo maintainer is active and accepting PR so this package is not dead just need some love from us the pakacge users.

I suggest that if we find something good that would be useful like how to "send the play/pause event" we will create a PR to add this functionality to everyone.

I have started with 2 PRs of adding Linux\Windows support and updating the example project and will probably add more on the way.

So let's make this package better by helping each other with the knowledge we gained.

**Comment by jonathantribouharet** [comment]
Hello,
no enum is needed as each code / action is specific to the application, this package is not dedicated to Youtube.
However if you find the code to work with Youtube it could be added in the documentation as an example.
Last time I've check (2 years ago), it was not easy to work with Youtube.

**Comment by guyluz11** [comment]
> Hello, no enum is needed as each code / action is specific to the application, this package is not dedicated to Youtube. However if you find the code to work with Youtube it could be added in the documentation as an example. Last time I've check (2 years ago), it was not easy to work with Youtube.

What about general settings like changing volume, does that depend on a specific app?

**Comment by jonathantribouharet** [comment]
Yes I'm ok with everything supported by the [official SDK](https://developers.google.com/cast).
Unfortunately the sender part (the one this lib used) is not open source and neither the cast protocol.
This project is based on another lib https://github.com/dylanmckay/gcast/tree/master (this guy has read chromium cast part to do this).

**Comment by guyluz11** [comment]
Ease functions I have added: Play/Pause (works only on YouTube), change volume (@Tiagocf2), Open Media (With easy call), getStatus function (to get current volume), and an example of opening PLEX.
Opening YouTube didn't work for me (stuck on the logo).
Added all of it to the example app
![image](https://github.com/jonathantribouharet/flutter_cast/assets/9304740/e2005a3c-74e2-44c8-a5a4-1b9a8dbb87dc)

This is on my repo but it will not work on iOS but does support Dart native because it is one or the other and I need Dart native more.
Will revert back to the iOS support and create a PR after my previous PRs get accepted to reduce complications.
Link for my fork if you want to test it out
https://github.com/guyluz11/flutter_cast/tree/multicast_version

For API I have looked on 
This one for the available messages 
https://developers.google.com/cast/docs/reference/web_receiver/cast.framework.messages

And this one for the properties of each message
https://developers.google.com/cast/docs/media/messages

Didn't check but I think it will be easier to use a project from JavaScript than Rust as a reference for the SDK
https://github.com/xat/castnow

**Comment by guyluz11** [comment]
Added function to open link in browser :D

And I have found a javascript package that manages to open YouTube videos 
https://github.com/i8beef/node-red-contrib-castv2/blob/master/lib/YouTubeController.js
I have started working on that and managed to get YouTube loungeToken and ok status from the last request but it still missing something.
Feel free to check it out, maybe you will see something that I have missed
https://github.com/guyluz11/flutter_cast/blob/09e845de9b872cb734e725b09b89d64a8fc281b9/lib/device.dart#L89
